File: DeforaOS/Apps/Web/DaPortal/system/xml.php
<?php //$Id$
//Copyright (c) 2010 Pierre Pronchery <khorben@defora.org>
//This file is part of DaPortal
//
//DaPortal is free software; you can redistribute it and/or modify
//it under the terms of the GNU General Public License version 2 as
//published by the Free Software Foundation.
//
//DaPortal is distributed in the hope that it will be useful,
//but WITHOUT ANY WARRANTY; without even the implied warranty of
//MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//GNU General Public License for more details.
//
//You should have received a copy of the GNU General Public License
//along with DaPortal; if not, write to the Free Software
//Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA



//private
//variables
global $xml_blacklisted, $xml_attrib_whitelist, $xml_tag_whitelist,
       $xml_content;
$xml_blacklisted = 1;
$xml_attrib_whitelist = array('alt', 'border', 'class', 'colspan', 'height',
		'href', 'size', 'src', 'style', 'title', 'width');
$xml_tag_whitelist = array('a', 'acronym', 'b', 'big', 'br', 'center', 'div',
		'font', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
		'hr', 'i', 'img', 'li', 'ol', 'p', 'pre', 'span', 'sub', 'sup',
		'table', 'tbody', 'td', 'th', 'tr', 'tt', 'u', 'ul');
$xml_content = '';


//functions
function _xml_text($content, &$error = '')
{
	global $xml_content;

	$xml_content = '';
	//FIXME factorize this code with validation
	$content = str_replace(array('<br>', '<hr>'), array('<br/>', '<hr/>'),
			$content);
	$content = str_replace(array('&lt;', '&gt;', '&quot;', '&'),
			array('<', '>', '"', '&amp;'), htmlentities($content));
	$content = preg_replace('/(<img [^>]*)>/', '\1/>', $content);
	$content = '<div>'.$content.'</div>';
	$parser = xml_parser_create(); //encoding should not matter
	xml_set_character_data_handler($parser, '_xml_text_update_data');
	if(($ret = xml_parse($parser, '<div>'.$content.'</div>', TRUE)) != 1)
		$error = xml_error_string(xml_get_error_code($parser))
			.' at line '.xml_get_current_line_number($parser)
			.', column '.xml_get_current_column_number($parser);
	xml_parser_free($parser);
	if($ret == 1)
		return str_replace(array('&amp;', '&lt;', '&gt;'),
				array('&', '<', '>'), $xml_content);
	$xml_content = '';
	return FALSE;
}

function _xml_text_update_data($parser, $data)
{
	global $xml_content;

	_warning($data, 0);
	$xml_content.=$data;
}


function _xml_validate($content, &$error = '')
{
	global $xml_blacklisted, $xml_error;

	$content = str_replace(array('<br>', '<hr>'), array('<br/>', '<hr/>'),
			$content);
	$content = str_replace(array('&lt;', '&gt;', '&quot;', '&'),
			array('<', '>', '"', '&amp;'), htmlentities($content));
	$content = preg_replace('/(<img [^>]*)>/', '\1/>', $content);
	$content = '<div>'.$content.'</div>';
	$parser = xml_parser_create(); //encoding should not matter
	if(xml_set_element_handler($parser, '_xml_validate_element_start',
				'_xml_validate_element_end') != TRUE)
	{
		xml_parser_free($parser);
		return FALSE;
	}
	$xml_blacklisted = 0;
	$xml_error = '';
	if(($ret = xml_parse($parser, $content, TRUE)) != 1)
		$error = xml_error_string(xml_get_error_code($parser))
			.' at line '.xml_get_current_line_number($parser)
			.', column '.xml_get_current_column_number($parser);
	xml_parser_free($parser);
	if($xml_blacklisted != 0 || strlen($xml_error) != 0)
	{
		$error = $xml_error;
		return FALSE;
	}
	$xml_blacklisted = 1;
	if($ret == 1)
		return TRUE;
	_error('XML error: '.$error, 0);
	return FALSE;
}

function _xml_validate_element_start($parser, $name, $attribs)
{
	global $xml_blacklisted, $xml_error, $xml_attrib_whitelist,
	       $xml_tag_whitelist;

	//return immediately if already detected as invalid
	//FIXME disable check in parser instead if possible
	if($xml_blacklisted != 0)
		return;
	//check the element name
	$wcnt = count($xml_tag_whitelist);
	for($i = 0; $i < $wcnt; $i++)
		if(strcasecmp($name, $xml_tag_whitelist[$i]) == 0)
			break;
	if($i == $wcnt) //tag not found
	{
		$xml_blacklisted = 1;
		$xml_error = 'The tag "'.$name.'" is forbidden';
		return;
	}
	//check every attribute
	$keys = array_keys($attribs);
	$attr = array();
	$i = 0;
	foreach($keys as $k)
		$attr[$i++] = $k;
	$acnt = count($attr);
	$wcnt = count($xml_attrib_whitelist);
	for($i = 0; $i < $acnt; $i++)
	{
		$a = $attr[$i];
		for($j = 0; $j < $wcnt; $j++)
			if(strcasecmp($a, $xml_attrib_whitelist[$j]) == 0)
				break;
		if($j == $wcnt) //attrib not found
			break;
	}
	if($i == $acnt)
		return;
	$xml_error = 'The attribute "'.$a.'" is forbidden';
	$xml_blacklisted = 1;
}

function _xml_validate_element_end($parser, $name)
{
}

?>

File: DeforaOS/System/src/libParser/tools/xml.c
/* $Id$ */
/* Copyright (c) 2010 Pierre Pronchery <khorben@defora.org> */
/* This file is part of DeforaOS System libParser */
/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. */
/* FIXME:
 * - plugin-based parsing system instead?
 *   * parser_new("XML", ...) ("Config", ...) */



#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include "System/Parser.h"


/* functions */
/* private */
/* error */
static int _error(char const * message, int ret)
{
	fputs("xml: ", stderr);
	perror(message);
	return ret;
}


/* usage */
static int _usage(void)
{
	fputs("Usage: xml [-w][-o output] filename\n"
"       xml -s [-w][-o output] string\n", stderr);
	return 1;
}


/* public */
/* main */
static void _main_node(XMLPrefs * prefs, FILE * fp, XMLNode * node,
		unsigned int level);

int main(int argc, char * argv[])
{
	int o;
	XMLPrefs prefs;
	XML * xml;
	XMLDocument * doc;
	char const * output = NULL;
	FILE * fp = stdout;
	char const * string = NULL;

	memset(&prefs, 0, sizeof(prefs));
	while((o = getopt(argc, argv, "o:s:w")) != -1)
		switch(o)
		{
			case 'o':
				output = optarg;
				break;
			case 's':
				string = optarg;
				break;
			case 'w':
				prefs.filters |= XML_FILTER_WHITESPACE;
				break;
			default:
				return _usage();
		}
	if(output != NULL && (fp = fopen(output, "w")) == NULL)
		return _error(output, 2);
	if(string == NULL)
	{
		if(optind + 1 != argc)
			return _usage();
		xml = xml_new(&prefs, argv[optind]);
	}
	else if(optind != argc)
		return _usage();
	else
		xml = xml_new_string(&prefs, string, strlen(string));
	if(xml == NULL)
		return error_print("xml");
	if((doc = xml_get_document(xml)) == NULL)
		error_print("xml");
	else
		_main_node(&prefs, fp, doc->root, 0);
	if(output != NULL)
		fclose(fp);
	xml_delete(xml);
	return 0;
}

static void _main_node(XMLPrefs * prefs, FILE * fp, XMLNode * node,
		unsigned int level)
{
	size_t i;

	if(node == NULL)
		return;
	switch(node->type)
	{
		case XML_NODE_TYPE_DATA:
			fwrite(node->data.buffer, 1, node->data.size, fp);
			break;
		case XML_NODE_TYPE_ENTITY:
			fprintf(fp, "&%s;", node->entity.name);
			break;
		case XML_NODE_TYPE_TAG:
			if(prefs->filters & XML_FILTER_WHITESPACE && level > 0)
			{
				fputc('\n', fp);
				for(i = 0; i < level; i++)
					fputs("  ", fp);
			}
			fprintf(fp, "<%s", node->tag.name);
			for(i = 0; i < node->tag.attributes_cnt; i++)
				fprintf(fp, " %s=\"%s\"",
						node->tag.attributes[i]->name,
						node->tag.attributes[i]->value);
			if(node->tag.childs_cnt == 0)
			{
				fputs("/>", fp);
				break;
			}
			fputs(">", fp);
			for(i = 0; i < node->tag.childs_cnt; i++)
				_main_node(prefs, fp, node->tag.childs[i],
						level + 1);
			if(prefs->filters & XML_FILTER_WHITESPACE && level == 0)
				fputc('\n', fp);
			fprintf(fp, "</%s>", node->tag.name);
			if(prefs->filters & XML_FILTER_WHITESPACE && level == 0)
				fputc('\n', fp);
			break;
	}
}
